Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,254 @@
+/// <reference path="../../definitions/node.d.ts" />
+/// <reference path="../../definitions/Q.d.ts" />
+
+
+import fs = require('fs');
+import os = require('os');
+import path = require('path');
+import tl = require('vsts-task-lib/task');
+import url = require('url');
+
+
+var win = os.type().match(/^Win/);
+tl.debug('win: ' + win);
+
+var repoRoot: string = tl.getVariable('build.sourcesDirectory');
+function makeAbsolute(normalizedPath: string): string {
+    tl.debug('makeAbsolute:' + normalizedPath);
+
+    var result = normalizedPath;
+    if (!path.isAbsolute(normalizedPath)) {
+        result = path.join(repoRoot, normalizedPath);
+        tl.debug('Relative file path: ' + normalizedPath + ' resolving to: ' + result);
+    }
+    return result;
+}
+
+function failTask(message: string) {
+    tl.setResult(tl.TaskResult.Failed, message);
+}
+
+// server endpoint
+var serverEndpoint = tl.getInput('serverEndpoint', true);
+var serverEndpointUrl : url.Url = url.parse(tl.getEndpointUrl(serverEndpoint, false));
+
+var serverEndpointAuth = tl.getEndpointAuthorization(serverEndpoint, false);
+var username = serverEndpointAuth['parameters']['username'];
+var password = serverEndpointAuth['parameters']['password'];
+
+// the root location which will be uploaded from
+var rootFolder: string = makeAbsolute(path.normalize(tl.getPathInput('rootFolder', true).trim()));
+if (!tl.exist(rootFolder)) {
+    failTask('The specified root folder: ' + rootFolder + ' does not exist.');
+}
+
+var clean: boolean = tl.getBoolInput('clean', true);
+var overwrite: boolean = tl.getBoolInput('overwrite', true);
+var flatten: boolean = tl.getBoolInput('flatten', true);
+
+function findFiles(): string[] {
+    tl.debug('Searching for files to upload');
+
+    var rootFolderStats = tl.stats(rootFolder);
+    if (rootFolderStats.isFile()) {
+        var file = rootFolder;
+        tl.debug(file + ' is a file. Ignoring all file patterns');
+        var parent = path.normalize(path.dirname(file));
+        return [parent, file];
+    }
+
+    var allFiles = tl.find(rootFolder);
+
+    // filePatterns is a multiline input containing glob patterns
+    var filePatterns: string[] = tl.getDelimitedInput('filePatterns', '\n', true);
+
+    for (var i = 0; i < filePatterns.length; i++) {
+        if (filePatterns[i] == '*') {
+            tl.debug('* matching everything, total: ' + allFiles);
+            return allFiles;
+        }
+    }
+
+    tl.debug('using: ' + filePatterns.length + ' filePatterns: ' + filePatterns + ' to search for files.');
+
+    // minimatch options
+    var matchOptions = { matchBase: true };
+    if (win) {
+        matchOptions["nocase"] = true;
+    }
+
+    tl.debug('Candidates found for match: ' + allFiles.length);
+    for (var i = 0; i < allFiles.length; i++) {
+        tl.debug('file: ' + allFiles[i]);
+    }
+
+    // use a set to avoid duplicates
+    var SortedSet = require('collections/sorted-set');
```

Comment:
Can this require('collections/sorted-set') be an import statement at the top of the file?